### PR TITLE
feat(overlay): native page reload dialog

### PR DESCRIPTION
**Pull request by Calinteodor**

### CODE REVIEW

File: react/features/app/components/AbstractApp.js [hunk 1]
```diff
@@ -73,23 +70,9 @@ export class AbstractApp extends BaseApp<Props, *> {
...
-     * @returns {ReactElement}
-     */
-    _createExtraElement() {
-        return (
-            <Fragment>
-                <OverlayContainer />
-            </Fragment>
-        );
-    }
+    _createExtraElement: () => ?React.ReactElement;
```
Comment (saghul) [inline react/features/app/components/AbstractApp.js:73]: We can just drop this altogether.

File: react/features/app/components/App.native.js [hunk 2]
```diff
@@ -149,6 +149,22 @@ export class App extends AbstractApp {
...
 
+    /**
+     * Creates an extra {@link ReactElement}s to be added (unconditionally)
+     * alongside the main element.
+     *
+     * @returns {ReactElement}
+     * @abstract
+     * @protected
+     */
+    _createExtraElement() {
```
Comment (saghul) [inline react/features/app/components/App.native.js:160]: Let's just drop this and have web add the overlay container, there is no need for the extra hook, this abstraction has already been painful enough...

File: react/features/overlay/components/OverlayContainer.js [hunk 3]
```diff
@@ -1,11 +1,8 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/components/OverlayContainer.js:1]: If you are removing the Flow annotations, pl make the file actual TS

File: react/features/overlay/components/_.native.js [hunk 4]
```diff
@@ -1,3 +1 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/components/_.native.js:1]: Why not drop this file altogether? There is no longer a use for it

File: react/features/overlay/components/index.js [hunk 5]
```diff
@@ -1,4 +1,3 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/components/index.js:1]: Let's drop the index file and put all components in a web directory

File: react/features/overlay/components/native/PageReloadOverlay.js [hunk 6]
```diff
@@ -79,15 +77,13 @@ class PageReloadOverlay extends AbstractPageReloadOverlay<Props> {
...
-            <OverlayFrame>
-                <ConfirmDialog
-                    cancelLabel = 'dialog.Cancel'
-                    confirmLabel = 'dialog.rejoinNow'
-                    descriptionKey = { `${t(message, { seconds: timeLeft })}` }
-                    onCancel = { this._onCancel }
-                    onSubmit = { this._onReloadNow }
-                    title = { title } />
-            </OverlayFrame>
+            <ConfirmDialog
```
Comment (saghul) [inline react/features/overlay/components/native/PageReloadOverlay.js:80]: Pl get rid of the overlay component altogether. This needs to be a standalone component.

File: react/features/overlay/components/native/index.js [hunk 7]
```diff
@@ -1,4 +1 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/components/native/index.js:1]: Drop this index too

File: react/features/overlay/components/web/OverlayFrame.js [hunk 8]
```diff
@@ -1,9 +1,5 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/components/web/OverlayFrame.js:1]: Ditto, TS.

File: react/features/overlay/index.js [hunk 9]
```diff
@@ -1,5 +1,3 @@
-// @flow
```
Comment (saghul) [inline react/features/overlay/index.js:1]: Drop this index file pl.

File: react/features/overlay/middleware.ts [hunk 10]
```diff
@@ -104,6 +109,8 @@ StateListenerRegistry.register(
             });
         }
 
+        dispatch(openDialog(PageReloadOverlay));
```
Comment (saghul) [inline react/features/overlay/middleware.ts:112]: This should only be done in RN, not it will also be dispatched in web.

File: react/features/overlay/middleware.ts [hunk 11]
```diff
@@ -104,6 +109,8 @@ StateListenerRegistry.register(
...
+
         if (NON_OVERLAY_ERRORS.indexOf(error.name) === -1 && typeof error.recoverable === 'undefined') {
```
Comment (saghul) [inline react/features/overlay/middleware.ts:114]: It's no longer necessary to set this fatal error, is it?

**Review by saghul — DISMISSED** [review]

File: react/features/overlay/components/native/PageReloadOverlay.js (hunk 6 above)
Comment (Calinteodor) [reply react/features/overlay/components/native/PageReloadOverlay.js:80]: This extends an Abstract component. We need its state object. I don't think it's a wise move.

File: react/features/overlay/middleware.ts (hunk 11 above)
Comment (Calinteodor) [reply react/features/overlay/middleware.ts:114]: Based on the logic that we have in AbstractPageReloadOverlay, I think it is.

**Comment by Calinteodor** [comment]
@saghul I applied what we discussed today, maybe you can check the PR. The problem right now is that I am getting this error and I can't figure out what I am missing? 

Maybe you can give me hint.

image

**Comment by Calinteodor** [comment]
@saghul I applied the changes that we discussed. You can take it for another spin. Thanks!

File: react/features/base/connection/functions.ts [hunk 12]
```diff
@@ -90,5 +90,5 @@ export function isInviteURLReady(stateOrGetState: IStateful): boolean {
  * {@code user@server.com}).
  */
 export function toJid(id: string, { authdomain, domain }: { authdomain?: string; domain?: string; }): string {
-    return id.indexOf('@') >= 0 ? id : `${id}@${authdomain || domain}`;
+    return id?.indexOf('@') >= 0 ? id : `${id}@${authdomain || domain}`;
```
Comment (saghul) [inline react/features/base/connection/functions.ts:93]: Unrelated change?

File: react/features/base/dialog/components/native/PageReloadDialog.tsx [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+        if (this.props.isNetworkFailure) {
+            title = 'dialog.conferenceDisconnectTitle';
+            message = 'dialog.conferenceDisconnectMsg';
+        } else {
+            title = 'dialog.conferenceReloadTitle';
+            message = 'dialog.conferenceReloadMsg';
+        }
+
+        this.state = {
+            message,
```
Comment (saghul) [inline react/features/base/dialog/components/native/PageReloadDialog.tsx:79]: This doesn't need to be part of the state. It can be just an instance variable.

File: react/features/base/dialog/components/native/PageReloadDialog.tsx [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+            timeLeft: timeoutSeconds,
+            // @ts-ignore
+            timeoutSeconds,
+            title
```
Comment (saghul) [inline react/features/base/dialog/components/native/PageReloadDialog.tsx:83]: Ditto.

File: react/features/base/dialog/components/native/PageReloadDialog.tsx [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    /**
+     * React Component method that executes once component is mounted.
+     *
+     * @inheritdoc
+     * @returns {void}
+     */
+    componentDidMount() {
+        if (typeof APP !== 'undefined') {
```
Comment (saghul) [inline react/features/base/dialog/components/native/PageReloadDialog.tsx:97]: Since this is used only in native this is not necessary.

File: react/features/base/dialog/components/native/PageReloadDialog.tsx [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+         * will be reloaded.
+         *
+         * @type {number}
+         */
+        const timeoutSeconds = 10 + randomInt(0, 20);
+
+        let message, title;
+
+        if (this.props.isNetworkFailure) {
+            title = 'dialog.conferenceDisconnectTitle';
```
Comment (saghul) [inline react/features/base/dialog/components/native/PageReloadDialog.tsx:71]: can't this be passed from mapStateToProps, actually?

File: react/features/overlay/actionTypes.ts [hunk 17]
```diff
@@ -25,14 +25,3 @@ export const MEDIA_PERMISSION_PROMPT_VISIBILITY_CHANGED
  * @public
  */
 export const SET_FATAL_ERROR = 'SET_FATAL_ERROR';
```
Comment (saghul) [inline react/features/overlay/actionTypes.ts:27]: Can't we also get rid of this?

File: react/features/overlay/actions.ts [hunk 18]
```diff
@@ -1,8 +1,10 @@
 import {
     MEDIA_PERMISSION_PROMPT_VISIBILITY_CHANGED,
-    SET_FATAL_ERROR,
-    SET_PAGE_RELOAD_OVERLAY_CANCELED
+    SET_FATAL_ERROR
```
Comment (saghul) [inline react/features/overlay/actions.ts:3]: Unused?

File: react/features/overlay/actions.ts [hunk 19]
```diff
@@ -26,37 +28,15 @@ export function mediaPermissionPromptVisibilityChanged(isVisible: boolean, brows
...
  */
-export function setFatalError(fatalError?: Object) {
-    return {
-        type: SET_FATAL_ERROR,
-        fatalError
-    };
-}
+export function fatalError(is?: boolean) {
+    if(is) {
+        openDialog(PageReloadDialog)
```
Comment (saghul) [inline react/features/overlay/actions.ts:38]: you need to dispatch this

File: react/features/overlay/components/web/AbstractPageReloadOverlay.js [hunk 20]
```diff
@@ -118,7 +100,7 @@ export default class AbstractPageReloadOverlay<P: Props>
      * @returns {boolean} - If this overlay needs to be rendered, {@code true};
      * {@code false}, otherwise.
      */
-    static needsRenderWeb(state: Object) {
+    static needsRender(state: Object) {
```
Comment (saghul) [inline react/features/overlay/components/web/AbstractPageReloadOverlay.js:103]: Isn't the logic here and the one in the new native dialog the same? would it make sense to have some `getFatalError(state)` helper perhaps?

File: react/features/overlay/components/web/OverlayContainer.tsx [hunk 21]
```diff
@@ -1,11 +1,10 @@
-// @flow
-
 import React, { Component } from 'react';
+import { ReactElementType } from 'react-window';
```
Comment (saghul) [inline react/features/overlay/components/web/OverlayContainer.tsx:2]: Weird import?

File: react/features/overlay/overlays.ts [hunk 22]
```diff
@@ -0,0 +1,19 @@
...
+import SuspendedOverlay from './components/web/SuspendedOverlay';
+// @ts-ignore
+import UserMediaPermissionsOverlay from './components/web/UserMediaPermissionsOverlay';
+
+/**
+ * Returns the list of available platform specific overlays.
+ *
+ * @returns {Array<Object>}
+ */
+export function getOverlays(): Array<Object> {
```
Comment (saghul) [inline react/features/overlay/overlays.ts:13]: This can now be inlined in the single place where it's used

File: react/features/overlay/middleware.ts [hunk 23]
```diff
@@ -102,10 +98,11 @@ StateListenerRegistry.register(
                 ...error,
                 ...getErrorExtraInfo(getState, error)
             });
-        }
-
-        if (NON_OVERLAY_ERRORS.indexOf(error.name) === -1 && typeof error.recoverable === 'undefined') {
```
Comment (saghul) [inline react/features/overlay/middleware.ts:107]: Why did you remove the check? Wouldn't this now show the reload overlay also for non-fatal errors?

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL!

File: react/features/base/connection/functions.ts (hunk 12 above)
Comment (Calinteodor) [reply react/features/base/connection/functions.ts:93]: This was giving me undefined, gonna check again.

File: react/features/overlay/actionTypes.ts (hunk 17 above)
Comment (Calinteodor) [reply react/features/overlay/actionTypes.ts:27]: Forgot about this one.

File: react/features/overlay/components/web/OverlayContainer.tsx (hunk 21 above)
Comment (Calinteodor) [reply react/features/overlay/components/web/OverlayContainer.tsx:2]: Replaced with any type.

File: react/features/overlay/middleware.ts (hunk 23 above)
Comment (Calinteodor) [reply react/features/overlay/middleware.ts:107]: I thought for those errors we don't show the reload overlay. Gonna revert this.

File: react/features/overlay/components/web/AbstractPageReloadOverlay.js (hunk 20 above)
Comment (Calinteodor) [reply react/features/overlay/components/web/AbstractPageReloadOverlay.js:103]: I think isErrorFatal(state) because it would return true or false based on if the error is fatal or not. The one from the native dialog checks the type of the error so it can update the reason string value.

Comment (saghul) [reply react/features/overlay/components/web/AbstractPageReloadOverlay.js:103]: I said get because in case it IS fatal you want the error object to extract the type etc, right? So it would help with both knowing if there is a fatal error and then extracting the info about it.

File: react/features/overlay/actions.ts (hunk 19 above)
Comment (Calinteodor) [reply react/features/overlay/actions.ts:38]: We have other actions that use openDialog and we dispatch them once. 
For ex. openLoginDialog.

**Comment by Calinteodor** [comment]
I am getting this weird error that I cannot find a way to tackle.
Any ideas?
image

File: react/features/overlay/actions.ts [hunk 24]
```diff
@@ -35,7 +35,7 @@ export function mediaPermissionPromptVisibilityChanged(isVisible: boolean, brows
  */
 export function fatalError(isFatal?: boolean) {
     if (isFatal) {
-        openDialog(PageReloadDialog);
+        return openDialog(PageReloadDialog);
```
Comment (Calinteodor) [inline react/features/overlay/actions.ts:38]: @saghul I think this makes the difference.

File: react/features/overlay/actions.web.ts [hunk 25]
```diff
@@ -0,0 +1,33 @@
...
+    };
+}
+
+/**
+ * Opens {@link PageReloadDialog}.
+ *
+ * @param {boolean} isFatal - If the value is true - we open PageReloadDialog.
+ * @returns {Function}
+ */
+export function fatalError(isFatal: boolean) { // eslint-disable-line @typescript-eslint/no-unused-vars
```
Comment (saghul) [inline react/features/overlay/actions.web.ts:31]: Ditto

File: react/features/overlay/functions.ts [hunk 26]
```diff
@@ -1,15 +1,27 @@
...
+import SuspendedOverlay from './components/web/SuspendedOverlay';
+// @ts-ignore
+import UserMediaPermissionsOverlay from './components/web/UserMediaPermissionsOverlay';
 /**
  * Returns the overlay to be currently rendered.
  *
  * @param {IReduxState} state - The Redux state.
  * @returns {?React$ComponentType<*>}
  */
 export function getOverlayToRender(state: IReduxState) {
```
Comment (saghul) [inline react/features/overlay/functions.ts:17]: Inline this wherever it need to be, functions.ts cannot import web stuff or we are going to break things...

**Review by saghul — CHANGES_REQUESTED** [review]
Left some comments, PTAL. Also, I could not see the dialog when putting the phone in airplane mode. It fails with "droppedError" which should trigger it.

```
2023-01-24 16:18:01.827560+0100 jitsi-meet[2068:540151] [JitsiMeetSDK] [modules/xmpp/strophe.util.js] Strophe: request id 16.3 error 0 happened
2023-01-24 16:18:01.827851+0100 jitsi-meet[2068:540479] [JitsiMeetSDK] [modules/xmpp/strophe.util.js] Strophe: request errored, status: 0, number of errors: 5
2023-01-24 16:18:01.827896+0100 jitsi-meet[2068:540479] [JitsiMeetSDK] [modules/xmpp/xmpp.js] (TIME) Strophe unknown:	 118179225.955041
2023-01-24 16:18:01.827937+0100 jitsi-meet[2068:540479] [JitsiMeetSDK] [modules/xmpp/strophe.ping.js] Ping interval cleared
2023-01-24 16:18:01.827982+0100 jitsi-meet[2068:540479] [JitsiMeetSDK] [modules/xmpp/xmpp.js] (TIME) Strophe disconnected:	 118179227.329791
2023-01-24 16:18:01.828042+0100 jitsi-meet[2068:540151] [JitsiMeetSDK] [modules/xmpp/xmpp.js] XMPP connection dropped!
2023-01-24 16:18:01.828219+0100 jitsi-meet[2068:540151] [JitsiMeetSDK] [modules/statistics/statistics.js] {"type":"operational","action":"connection.failed","attributes":{"error_type":"connection.droppedError","error_message":"connection-dropped-error","suspend_time":16,"time_since_last_success":10355}}
2023-01-24 16:18:01.939728+0100 jitsi-meet[2068:540479] [JitsiMeetSDK] rn-webrtc:pc:DEBUG 0 close +2s
2023-01-24 16:18:01.951870+0100 jitsi-meet[2068:534066] [/Users/saghul/work/jitsi/jitsi-meet/ios/app/src/ViewController.m:44] JitsiMeetViewDelegate CONFERENCE_TERMINATED {
    error = "connection.droppedError";
    isAudioMuted = 0;
    url = "https://meet.jit.si/saghul";
}
```

File: react/features/base/lib-jitsi-meet/functions.web.ts [hunk 27]
```diff
@@ -10,5 +10,6 @@ export async function loadConfig(url: string) { // eslint-disable-line @typescri
     // Return "the config.js file" from the global scope - that is how the
     // Web app on both the client and the server was implemented before the
     // React Native app was even conceived.
+    // @ts-ignore
```
Comment (saghul) [inline react/features/base/lib-jitsi-meet/functions.web.ts:13]: This should be unnecessary when using tsconfig.web.json as your tsconfig.

File: react/features/conference/components/web/Conference.js [hunk 28]
```diff
@@ -234,7 +242,7 @@ class Conference extends AbstractConference<Props, *> {
 
                     { _showPrejoin || _showLobby || <Toolbox /> }
 
-                    {_notificationsVisible && (_overflowDrawer
+                    {_notificationsVisible && !_isAnyOverlayVisible && (_overflowDrawer
```
Comment (saghul) [inline react/features/conference/components/web/Conference.js:245]: Why is this necessary now?

Comment (Calinteodor) [reply react/features/conference/components/web/Conference.js:245]: So notifications don't display overlay the overlays.

File: react/features/overlay/middleware.ts [hunk 29]
```diff
@@ -92,20 +94,27 @@ StateListenerRegistry.register(
...
+        // eslint-disable-next-line no-negated-condition
         if (typeof APP !== 'undefined') {
             APP.API.notifyError({
                 ...error,
-                ...getErrorExtraInfo(getState, error)
+                ...getErrorExtraInfo(state, error)
             });
+        } else {
+            // @ts-ignore
+            store.dispatch(openPageReloadDialog());
```
Comment (saghul) [inline react/features/overlay/middleware.ts:112]: this is already done below, you can drop this else branch since we only want to show the page reload dialog in case of an unrecoverable error

**Review by saghul — APPROVED** [review]
:fire:
